libraries/AP_MSP/AP_MSP_Telem_DJI.cpp

```diff
@@ -121,4 +124,120 @@ MSPCommandResult AP_MSP_Telem_DJI::msp_process_out_esc_sensor_data(sbuf_t *dst)
     return MSP_RESULT_ACK;
 }
 
+void AP_MSP_Telem_DJI::update_home_pos(home_state_t &home_state)
+{
+    AP_MSP_Telem_Backend::update_home_pos(home_state);
+#if OSD_ENABLED
+    AP_MSP *msp = AP::msp();
+    if (msp == nullptr) {
+        return;
+    }
+    AP_OSD *osd = AP::osd();
+    if (osd == nullptr) {
+        return;
+    }
+    if (!osd_stats_enabled()) {
+        return;
+    }
+    WITH_SEMAPHORE(osd->get_semaphore());
+    // override telemetry with max distance and altitude info
+    // alternate max distance with traveled distance every 2 seconds
+    if (msp->_msp_status.slow_flashing_on) {
+        home_state.home_distance_m = osd->get_stats_info().max_dist_m;
+    } else {
+        home_state.home_distance_m = osd->get_stats_info().last_distance_m;
+    }
+    home_state.rel_altitude_cm = osd->get_stats_info().max_alt_m * 100;
+#endif    
+}
+
+void AP_MSP_Telem_DJI::update_battery_state(battery_state_t &_battery_state)
+{
+    AP_MSP_Telem_Backend::update_battery_state(_battery_state);
+#if OSD_ENABLED
+    AP_OSD *osd = AP::osd();
+    if (osd == nullptr) {
+        return;
+    }
+    if (!osd_stats_enabled()) {
+        return;
+    }
+    WITH_SEMAPHORE(osd->get_semaphore());
+    // override telemetry with max current and voltage info
+    _battery_state.batt_current_a = osd->get_stats_info().max_current_a;
+    _battery_state.batt_voltage_v = osd->get_stats_info().min_voltage_v;
+#endif    
+}
+
+void AP_MSP_Telem_DJI::update_gps_state(gps_state_t &gps_state)
+{
+    AP_MSP_Telem_Backend::update_gps_state(gps_state);
+#if OSD_ENABLED
+    AP_OSD *osd = AP::osd();
+    if (osd == nullptr) {
+        return;
+    }
+    if (!osd_stats_enabled()) {
+        return;
+    }
+    WITH_SEMAPHORE(osd->get_semaphore());
+    // override telemetry with max speed info
+    gps_state.speed_cms = osd->get_stats_info().max_speed_mps * 100;
+#endif    
+}
+
+void AP_MSP_Telem_DJI::update_flight_mode_str(char *flight_mode_str, bool wind_enabled)
+{
+    AP_MSP_Telem_Backend::update_flight_mode_str(flight_mode_str, wind_enabled);
+#if OSD_ENABLED
+    AP_OSD *osd = AP::osd();
+    if (osd == nullptr) {
```

Comment:
Are we supposed to be taking a semaphore on this?  If not (probably done by caller...) we don't need this pointer here.

Reply:
mhmm...we read atomic stuff, like bools or <= 32 bit data so no need for a semaphore, but we need the pointer further down, not sure I understand your objection here

Reply:
Uh - looks like you're taking a semaphore instead?!

Reply:
there was a stale unused osd* in the overriden AP_MSP_Telem_DJI::update_flight_mode_str() which I removed, but here I need the pointer to access osd->screen[], I'm not locking on osd->get_semaphore()